### PR TITLE
[native pos] Remove busy-waiting-loop to reduce the CPU usage

### PR DESCRIPTION
**Pull request by miaoever**
The busy-waiting-loop in PrestoSparkNativeTaskExecutorFactory is running in spark executor main thread during task execution which consuming up to 40% of spark executor JVM CPU in our local profiling. In order to reduce the CPU usage, We make the task main thread waiting on two signals (task finish and output is ready) rather than busy waiting. We also removed unused field in the same class (functionAndTypeManager).

As the results, in our test run in cluster, we're seeing the same query using 30% less CPU time in total query CPU usage.

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

File: presto-spark-base/src/test/java/com/facebook/presto/spark/execution/http/TestPrestoSparkHttpClient.java [hunk 1]
```diff
@@ -699,6 +742,35 @@ public void testInfoFetcherWithRetry()
...
+        taskInfoFetcher.start();
+        try {
+            synchronized (lock) {
+                while (!isTaskDone(taskInfoFetcher.getTaskInfo())) {
+                    lock.wait();
+                }
+            }
+        }
+        catch (InterruptedException e) {
+            e.printStackTrace();
```
Comment (mbasmanova) [inline presto-spark-base/src/test/java/com/facebook/presto/spark/execution/http/TestPrestoSparkHttpClient.java:763]: Do we need this?

**Review by mbasmanova — APPROVED** [review]
@miaoever Thanks.

@shrinidhijoshi Would you take a look as well?

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkNativeTaskExecutorFactory.java [hunk 2]
```diff
@@ -425,74 +422,73 @@ public boolean hasNext()
...
-            // While the remote task is still running and there is no output pages,
-            // we need to simulate a busy-loop to avoid returning null.
-            while (true) {
-                try {
-                    // For ShuffleMap Task, this will always return Optional.empty()
-                    Optional<SerializedPage> pageOptional = nativeExecutionTask.pollResult();
-
-                    if (pageOptional.isPresent()) {
-                        return pageOptional;
+            try {
```
Comment (shrinidhijoshi) [inline presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkNativeTaskExecutorFactory.java:456]: I wonder if we can simplify this code by writing 2 different iterators - 1. shuffleMap task and 2. for result task, 
Reason being 
- Behavior of Shuffle Map task is fundamentally different and is more straight forward (wait for `CompletableFuture` returned by `NativeExecutionTask`) Most of the stage in all of our prod workload would use the ShuffleMap iterator as that is the 99% use-case. 

I am worried the multiple `wait()/notifyAll()` pattern here might be hiding behaviors that are even harder to spot than the current busy-wait bug

**Review by shrinidhijoshi — APPROVED** [review]
LGTM! Thanks for investigating this @miaoever .

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkNativeTaskExecutorFactory.java (hunk 2 above)
Comment (miaoever) [reply presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkNativeTaskExecutorFactory.java:456]: Thanks for the reviews @shrinidhijoshi. 

Technically waiting for `CompletableFuture` (`CompletableFuture::get()`) is also a busy-waiting which is a not good idea to have it on the main thread. 

On the other hand, for our scenarios - asynchronous communication between different threads/processes, I believe the `wait()/notifyAll()` is the right java primitives to use if we can't use more built-in high level blocks (e.g. blockingQueue etc.)

File: presto-spark-base/src/test/java/com/facebook/presto/spark/execution/http/TestPrestoSparkHttpClient.java (hunk 1 above)
Comment (miaoever) [reply presto-spark-base/src/test/java/com/facebook/presto/spark/execution/http/TestPrestoSparkHttpClient.java:763]: Removed.

**Pull request merged**